pkg/controller/petset/identity_mappers.go

```diff
@@ -0,0 +1,202 @@
+package petset
+
+import (
+	"crypto/md5"
+	"fmt"
+	"sort"
+	"strings"
+
+	"github.com/golang/glog"
+	"k8s.io/kubernetes/pkg/api"
+	api_pod "k8s.io/kubernetes/pkg/api/pod"
+	"k8s.io/kubernetes/pkg/apis/apps"
+	"k8s.io/kubernetes/pkg/util/sets"
+)
+
+const svcLocalSuffix = "svc.cluster.local"
+
+// identityMapper is an interface for assigning identities to a pet.
+// All existing identity mappers just append "-(index)" to the petset name to
+// generate a unique identity. This is used in claims/DNS/hostname/petname
+// etc. There's a more elegant way to achieve this mapping, but we're
+// taking the simplest route till we have data on weather users will need
+// more customization.
+type identityMapper interface {
+	AddIdentity(id string, pet *api.Pod)
+	GetIdentity(pod *api.Pod) string
+}
+
+func NewIdentityMappers(ps *apps.PetSet) []identityMapper {
+	return []identityMapper{
+		&NameIdentityMapper{ps},
+		&NetworkIdentityMapper{ps},
+		&VolumeIdentityMapper{ps},
+	}
+}
+
+type NetworkIdentityMapper struct {
+	ps *apps.PetSet
+}
+
+func (n *NetworkIdentityMapper) AddIdentity(id string, pet *api.Pod) {
+	pet.Annotations[api_pod.PodHostnameAnnotation] = fmt.Sprintf("%v-%v", n.ps.Name, id)
+	pet.Annotations[api_pod.PodSubdomainAnnotation] = n.ps.Spec.ServiceName
+	return
+}
+
+func (n *NetworkIdentityMapper) GetIdentity(pet *api.Pod) string {
+	return n.String(pet)
+}
+
+func (n *NetworkIdentityMapper) String(pet *api.Pod) string {
+	hostname := pet.Annotations[api_pod.PodHostnameAnnotation]
+	subdomain := pet.Annotations[api_pod.PodSubdomainAnnotation]
+	return fmt.Sprintf("%v.%v.%v.%v", hostname, subdomain, n.ps.Namespace, svcLocalSuffix)
+}
+
+type VolumeIdentityMapper struct {
+	ps *apps.PetSet
+}
+
+func (v *VolumeIdentityMapper) AddIdentity(id string, pet *api.Pod) {
+	petVolumes := []api.Volume{}
+	petClaims := GetClaims(id, v.ps)
+
+	// These volumes will all go down with the pod. If a name matches one of
+	// the claims in the pet set, it gets clobbered.
+	podVolumes := map[string]api.Volume{}
+	for _, podVol := range pet.Spec.Volumes {
+		podVolumes[podVol.Name] = podVol
+	}
+
+	// Insert claims for the idempotent petSet volumes
+	for name, claim := range petClaims {
```

Comment:
Clarify? It's 2 lists into 2 maps and I'm looking for a key match, so last one should always win? 
i.e if i have 2 templates with the same name, like:
```
  volumeClaimTemplates:
  - metadata:
      name: datadir
  - metadata:
      name: datadir
      annotations:
        volume.alpha.kubernetes.io/storage-class: anything
```

the second one makes it in because we're iterating the list into a map.
If we have:
```
  petSpec:
      volumes:
      - name: datadir
        mountPath: /foo
      - name: datadir
         mountPath: /bar
       - name: datadir
         hostPath

  volumeClaimTemplates:
  - metadata:
      name: datadir
```

The last one (host path) will get clobbered. This is what we should validate against. 
Or are you asking for different behavior? 
Maps make the comparison less convoluted, because I'm modifying the claim name to fit the id of the pet, but matching the spec.Volumes on the pre-modified name.